### DISCUSSION TITLE
Add shebang to executables?

### DISCUSSION THREAD

**Discussion by mpartel** [emscripten-core/emscripten#21059] [Ideas]
I was wondering, why not add `#!/usr/bin/env node` or something to all built executables (and `chmod +x` them)?

I don't see a downside, at least if this is optional/configurable, and this would be an easy fix for #17008/#7086: i.e. [build-system self execution](https://emscripten.org/docs/compiling/Building-Projects.html#build-system-self-execution).

(Perhaps this is [already possible](https://github.com/mesonbuild/meson/issues/6669#issuecomment-589893915), but I did not find a way to enable it.)

**Comment by mpartel** [comment]
Ok, not as easy as I thought. Such build tools need `NODERAWFS` to work (unless they're pure stdio), but the project as a whole might not want `NODERAWFS`.

So instead of just running with `node`, we'd probably need some emscripten-specific wrapper that injects `NODERAWFS` support before invoking `node`. I have no idea how difficult that would be.

Anyway, I did get Freetype (#17008) to build with this workaround:
```bash
emconfigure ./configure
env EMCC_CFLAGS=-sNODERAWFS=1 emmake make "$(pwd)/objs/apinames"
sed -i '1s;^;\#\!\/usr\/bin\/env node\n;' objs/apinames
chmod +x objs/apinames
emmake make
```

**Comment by sbc100** [comment]
I like this idea, and indeed we already do do this in some case.  See https://github.com/emscripten-core/emscripten/blob/49ab23ad9f636377eb1ee227445b95040da4afdb/tools/link.py#L496-L500.

Right now we only do this under certain circumstances, and we don't expose it as a user-facing option yet: https://github.com/emscripten-core/emscripten/blob/49ab23ad9f636377eb1ee227445b95040da4afdb/tools/link.py#L659-L666

**Comment by vadimkantorov** [comment]
I'm interested to run the Emscripten-produced JS/wasm with node to do CI-testing (e.g. I managed to compile Perl to wasm and would like to run some `./perl -e 'print("hello world\n");'`).

What are the necessary emcc/linker flags to be able to direcly run the produced executable with `node` without resorting to any node-specific loaders and so forth?

@sbc100 Are there any hacks to auto-add this shebang? Or are there any other options besides `-s NODERAWFS=1` and sed like proposed above? or should I also add `-s ENVIRONMENT=node`?

**Reply by sbc100** [reply]
`node` is already part of the default `-sENVIRONMENT` but if you want to save on code size you can limit that output to only running on node using `-s ENVIRONMENT=node` yes (assuming you don't want to run on the web).
